### PR TITLE
cardano-cli:  factor

### PR DESCRIPTION
**Pull request by deepfire**
1. Factor `runCommand` into components
1. New modules:
   - `Cardano.CLI.Delegation`
1. Small readability improvements
1. Mass-apply the https://github.com/input-output-hk/cardano-node/pull/164#pullrequestreview-284571398 to the CLI code.
1. Add in-line documentation for exported functions in CLI code.

### CODE REVIEW

File: cardano-node/src/Cardano/CLI/Genesis.hs [hunk 1]
```diff
@@ -21,17 +32,102 @@ import           System.Posix.Files (ownerReadMode, setFileMode)
...
+newtype NewDirectory =
+  NewDirectory FilePath
+  deriving (Eq, Ord, Show, IsString)
+
+newtype GenesisFile =
+  GenesisFile FilePath
+  deriving (Eq, Ord, Show, IsString)
+
+-- | Parameters required for generation of new genesis.
+data GenesisParameters where
```
Comment (erikd) [inline cardano-node/src/Cardano/CLI/Genesis.hs]: Does that need to be a `GADT` style `data` decl?

File: cardano-node/src/Cardano/CLI/Delegation.hs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE RecordWildCards            #-}
```
Comment (erikd) [inline cardano-node/src/Cardano/CLI/Delegation.hs]: The accursed `RecordWildCards`!

File: cardano-node/src/Cardano/CLI/Key.hs [hunk 3]
```diff
@@ -52,20 +60,69 @@ newtype NewVerificationKeyFile =
...
+    loop :: IO ByteString
+    loop = do
+      (v1, v2) <- (,) <$> readOne prompt <*> readOne "Repeat to validate: "
+      if v1 == v2
+        then pure v1
+        else hPutStrLn stdout ("Sorry, entered passwords don't match." :: String)
+             >> loop
+
+    readOne :: String -> IO ByteString
+    readOne pr = do
```
Comment (erikd) [inline cardano-node/src/Cardano/CLI/Key.hs]: :tada:

File: cardano-node/src/Cardano/CLI/Tx.hs [hunk 4]
```diff
@@ -112,3 +125,110 @@ genesisUTxOTxIn gc genSk genAddr =
...
+-- | Submit a transaction to a node specified by topology info.
+nodeSubmitTx
+  :: CLIOps IO
+  -> TopologyInfo
+  -> CardanoConfiguration
+  -> GenTx (ByronBlockOrEBB ByronConfig)
+  -> IO ()
+nodeSubmitTx co topology cc tx =
+  withRealPBFT co cc $
+    \p@(ProtocolRealPBFT _ _ _ _ _) -> do
```
Comment (erikd) [inline cardano-node/src/Cardano/CLI/Tx.hs]: That can be written as:
```
    \p@ProtocolRealPBFT {} -> do
```

File: cardano-node/src/Cardano/CLI/Genesis.hs (hunk 1 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Genesis.hs]: Done.

File: cardano-node/src/Cardano/CLI/Tx.hs (hunk 4 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Tx.hs]: Done!

File: cardano-node/app/cardano-cli.hs [hunk 5]
```diff
@@ -31,14 +29,16 @@ import           Cardano.Crypto ( AProtocolMagic(..)
...
+import           Cardano.CLI.Key
 import           Cardano.CLI.Ops (decideCLIOps)
 import           Cardano.CLI.Run
 
 main :: IO ()
 main = do
-  CLI {mainCommand, protocol} <- Opt.customExecParser pref opts
-  ops <- decideCLIOps protocol
-  catchIO (runCommand ops mainCommand)
+  co <- Opt.customExecParser pref opts
```
Comment (Jimbo4350) [inline cardano-node/app/cardano-cli.hs]: :100:

File: cardano-node/src/Cardano/CLI/Delegation.hs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      ( Dlg.issuerVK cert) issuerVK'
+    | Dlg.issuerVK cert /= issuerVK'
+    ]
+  , [ F.sformat ("Certificate delegate ".vk." doesn't match expected: ".vk)
+      ( Dlg.delegateVK cert) delegateVK'
+    | Dlg.delegateVK cert /= delegateVK'
+    ]
+  ]
+  where
+    magic' = Annotated magic (serialize' magic)
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Delegation.hs]: Type signatures :stuck_out_tongue:

File: cardano-node/src/Cardano/CLI/Key.hs [hunk 7]
```diff
@@ -1,23 +1,27 @@
...
 module Cardano.CLI.Key
   ( SigningKeyFile(..)
   , NewSigningKeyFile(..)
   , VerificationKeyFile(..)
   , NewVerificationKeyFile(..)
-  , prettySigningKeyPub
+  , prettyPublicKey
   , readSigningKey
   , readVerificationKey
+  , PasswordRequirement(..)
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Key.hs]: Did we decide on how to arrange exports in the style doc?

File: cardano-node/src/Cardano/CLI/Key.hs [hunk 8]
```diff
@@ -52,20 +59,69 @@ newtype NewVerificationKeyFile =
...
+  =  GetPassword
+  |  EmptyPassword
+  deriving (Eq, Show)
+
+type PasswordPrompt = String
 
-prettySigningKeyPub :: SigningKey -> Text
-prettySigningKeyPub (Crypto.toVerification -> vk) = TL.toStrict
-  $  "public key hash: " <> (F.format Crypto.hashHexF . CC.Common.addressHash $ vk) <> "\n"
+-- | Print some invariant properties of a public key:
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Key.hs]: :+1:

File: cardano-node/src/Cardano/CLI/Tx.hs [hunk 9]
```diff
@@ -1,114 +1,234 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE RankNTypes #-}
 {-# OPTIONS_GHC -Wno-all-missed-specialisations #-}
 
 module Cardano.CLI.Tx
-  ( prettyAddress
-  , txSpendUTxOByronPBFT
+  ( TxFile(..)
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Tx.hs]: I now see where you were going with this :+1:

File: cardano-node/src/Cardano/CLI/Tx.hs [hunk 10]
```diff
@@ -1,114 +1,234 @@
...
     pmid
     Crypto.SignTx
     sk
-    (CC.UTxO.TxSigData txid))
+    (UTxO.TxSigData txid))
 
 -- | Given a genesis, and a pair of signing key and address, reconstruct a TxIn
 --   corresponding to the genesis UTxO entry.
-genesisUTxOTxIn :: CC.Genesis.Config -> SigningKey -> Address -> CC.UTxO.TxIn
+genesisUTxOTxIn :: Genesis.Config -> SigningKey -> Common.Address -> UTxO.TxIn
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Tx.hs]: :+1:

File: cardano-node/src/Cardano/CLI/Tx/Submission.hs [hunk 11]
```diff
@@ -2,7 +2,6 @@
 {-# LANGUAGE DataKinds           #-}
 {-# LANGUAGE GADTs               #-}
 {-# LANGUAGE RankNTypes          #-}
-{-# LANGUAGE NamedFieldPuns      #-}
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Tx/Submission.hs]: :100:

File: cardano-node/src/Cardano/CLI/Tx/Submission.hs [hunk 12]
```diff
@@ -98,13 +96,13 @@ submitTx :: ( RunDemo blk
...
-submitTx CardanoConfiguration{ccSocketPath} pInfoConfig nodeId tx tracer = do
-    socketDir <- canonicalizePath =<< makeAbsolute ccSocketPath
+submitTx cc protoInfoConfig nodeId tx tracer = do
+    socketDir <- canonicalizePath =<< makeAbsolute (ccSocketPath cc)
     let addr = localSocketAddrInfo (socketDir </> localSocketFilePath nodeId)
     connectTo
       nullTracer
       (,)
-      (localInitiatorNetworkApplication tracer pInfoConfig tx)
+      (localInitiatorNetworkApplication tracer protoInfoConfig tx)
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Tx/Submission.hs]: `protoInfoConfig` > `pInfoConfig` :+1:

File: cardano-node/src/Cardano/Common/Protocol.hs [hunk 13]
```diff
@@ -53,44 +55,54 @@ data Protocol =
...
-            } = ccCore
+            } = ccCore cc
         genHash = either (throw . ConfigurationError) identity $
                   decodeAbstractHash coGenesisHash
         cvtRNM :: RequireNetworkMagic -> RequiresNetworkMagic
         cvtRNM NoRequireNetworkMagic = RequiresNoMagic
         cvtRNM RequireNetworkMagic   = RequiresMagic
 
-    gcE <- runExceptT (Genesis.mkConfigFromFile (cvtRNM $ coRequiresNetworkMagic ccCore) coGenesisFile genHash)
+    gcE <- runExceptT (Genesis.mkConfigFromFile
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Common/Protocol.hs]: Ah these are going to disrupt changes I'm currently making here re: exception handling. That fine though, I'll deal with it in the rebase :).

Comment (deepfire) [reply cardano-node/src/Cardano/Common/Protocol.hs]: Dropping NFP/RWC was kind of a global change..  I hope you're not too angry. : -/

File: cardano-node/src/Cardano/CLI/Key.hs (hunk 7 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Key.hs]: I've added grouping to underscore the intent -- within each group the types precede functions.

Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Key.hs]: A relevant discussion in the style document: https://github.com/input-output-hk/cardano-node/pull/164#discussion_r321786601

File: cardano-node/src/Cardano/Common/Protocol.hs (hunk 13 above)
Comment (Jimbo4350) [reply cardano-node/src/Cardano/Common/Protocol.hs]: Just the opposite in fact :slightly_smiling_face: Happy to see them go!

File: cardano-node/src/Cardano/CLI/Delegation.hs (hunk 6 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Delegation.hs]: Well, it's not that they are required by the style guide in `where` or `let` clauses?

Comment (Jimbo4350) [reply cardano-node/src/Cardano/CLI/Delegation.hs]: > is it really true that type signatures would add significant value here?

It's not life or death but it removes ambiguity for me. I don't see in this case how being more explicit would hurt understandability. Don't let this comment stop you from merging.

File: cardano-node/src/Cardano/CLI/Tx.hs (hunk 10 above)
Comment (erikd) [reply cardano-node/src/Cardano/CLI/Tx.hs]: This might also be something to look at for refactoring.

Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Tx.hs]: Done, no more `CC.` in cardano-node:
```
$ fgrep 'CC.' . -rI | wc -l
0
```

Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Tx.hs]: Done, no more `CC.` in cardano-node:
```
[nix-shell:~/cardano-node]$ fgrep 'CC.' . -rI | wc -l
0
```

Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Tx.hs]: Done!

File: cardano-node/src/Cardano/CLI/Delegation.hs (hunk 6 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Delegation.hs]: Done!

File: cardano-node/app/cardano-cli.hs [hunk 14]
```diff
@@ -76,38 +76,40 @@ parseClientCommand =
...
 
 -- | Values required to create genesis.
+parseGenesisParameters :: Parser GenesisParameters
+parseGenesisParameters =
+  GenesisParameters
+  <$> parseUTCTime
+      "start-time"
+      "Start time of the new cluster to be enshrined in the new genesis."
+  <*> parseFilePath
+      "protocol-parameters-file"
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: Done.

File: cardano-node/src/Cardano/CLI/Genesis.hs [hunk 15]
```diff
@@ -21,42 +31,132 @@ import           System.Posix.Files (ownerReadMode, setFileMode)
...
+        True                -- Not using delegate keys unsupported.
+
+  genesisSpec <- case mGenesisSpec of
+    Left e -> throwIO $ GenesisSpecError e
+    Right x -> pure x
+
+  mGData <- runExceptT $ Genesis.generateGenesisData (gpStartTime gp) genesisSpec
+  case mGData of
+    Left e -> throwIO $ GenesisGenerationError e
+    Right x -> pure x
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Genesis.hs]: I was looking at it, agonisingly, myself..  Problem is, it use a lot of local bindings, which will make the contact surface fairly large.

Moreover, my gut instinct is that this surface may yet change.  That's why I kept it as is, thus far.

But will do.

[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Genesis.hs]: Went to the town and split / redid it with `ExceptT`.

Not sure I was successful at extracting the pure part, though..

File: cardano-node/app/cardano-cli.hs [hunk 16]
```diff
@@ -31,14 +29,16 @@ import           Cardano.Crypto ( AProtocolMagic(..)
...
 import           Cardano.CLI.Run
 
 main :: IO ()
 main = do
-  CLI {mainCommand, protocol} <- Opt.customExecParser pref opts
-  ops <- decideCLIOps protocol
-  catchIO (runCommand ops mainCommand)
+  co <- Opt.customExecParser pref opts
+  ops <- decideCLIOps (protocol co)
+  catchIO (runCommand ops (mainCommand co))
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: Yes, and it looks like less text too.

But also we lose the property of having a `let`-like construct that underscores the genesis of names / centralises them.

**Comment by erikd** [comment]
Thins thing can be merged!

bors r+

**Pull request merged**